File: 프로그래머스/lv2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    let visited = Array.from({length:dungeons.length},()=>false);
        let answer = 0;
    
    const dfs = (h,c)=>{
        for(let i=0; i<dungeons.length; i++){
            if(!(visited[i]) && dungeons[i][0] <= h){
            visited[i] = true;
            dfs(h-dungeons[i][1] ,c+1);
            visited[i] = false;
            }
        }
        answer = Math.max(answer,c);
    }
    dfs(k,0);
    return answer
}

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.Arrays;

class Solution {
    
    int answer = 0;
    boolean[] visited;
    int[][] dungeon;

    public int solution(int k, int[][] dungeons) {

        visited = new boolean[dungeons.length];
        Arrays.fill(visited, false);
        dungeon = dungeons;

        dfs(k, 0);
        
        return answer;
    }

    private void dfs(int k, int count) {
        for (int i = 0; i <dungeon.length; i++) {
            if (!visited[i] && dungeon[i][0] <= k) {
                visited[i] = true;
                dfs(k-dungeon[i][1],count+1);
                visited[i] = false;
            }
        }
        answer = Math.max(count, answer);
    }
}